app/agent/runtime/qemu/qemu.service.ts

```diff
@@ -0,0 +1,542 @@
+  /**
+   * If the instance files are on disk, returns the result object.
+   * If the instance does not exist on disk, returns null.
+   * */
+  async getRuntimeInfo(): Promise<{ status: "on" | "off" | "paused"; info: VmInfo } | null> {
+    const info = await this.readVmInfoFile().catch((err) => {
+      if (err?.code === "ENOENT") {
+        return null;
+      }
+      throw err;
+    });
+    if (info === null) {
+      return null;
+    }
+
+    let sock: Socket;
+    try {
+      sock = await this.qmpConnect();
+    } catch (err: any) {
+      if (err instanceof QmpConnectionFailedError) return null;
+      throw err;
+    }
+
+    await this.sendQMPMessage(sock, "query-status");
+    const m = await this.waitForQMPMessageWithTimeout(sock);
+    if (m === void 0) {
+      return null;
+    }
+    sock.destroy();
+    return {
+      status: match((m as any).return.status)
+        .with("running", () => "on" as const)
+        .with("suspended", () => "paused" as const)
+        .with("paused", () => "paused" as const)
+        .with("postmigrate", () => "paused" as const)
+        .with("shutdown", () => "off" as const)
+        .with("internal-error", () => "off" as const)
+        .with("io-error", () => "off" as const)
+        .with(null, () => "off" as const)
+        .exhaustive(),
+      info,
+    };
+  }
+
+  async withRuntime<T>(
+    config: {
+      /**
+       * SSH Configuration for connecting to the VM
+       */
+      ssh: Omit<SSHConfig, "host">;
+      /**
+       * Optional kernel for the runtime
+       * Containers won't support changing the kernel
+       */
+      kernelPath?: string;
+      /**
+       * How should the final filesystem look like?
+       * This is a map in the form <RUNTIME_PATH> -> <HOST_SPEC>
+       * An entry for / is required
+       */
+      fs: Record<
+        string,
+        { device: string; readonly: boolean } | { mountPoint: string; readonly: boolean }
+      >;
+      /**
+       * Defaults to true.
+       */
+      shouldPoweroff?: boolean;
+      /** Defaults to true. */
+      cleanupAfterStop?: boolean;
+      memSizeMib: number;
+      vcpuCount: number;
+    },
+    fn: (args: {
+      ssh: NodeSSH;
+      sshConfig: SSHConfig;
+      runtimePid: number;
+      vmIp: string;
+      ipBlockId: IpBlockId;
+    }) => Promise<T>,
+  ): Promise<T> {
+    const t1 = performance.now();
+    const kernelPath = config.kernelPath ?? this.agentConfig.defaultKernel;
+    const shouldPoweroff = config.shouldPoweroff ?? true;
+    let vmStarted = false;
+    if (config.fs["/"] === void 0) {
+      throw new Error("No root fs specified");
+    }
+    for (const path of Object.keys(config.fs)) {
+      if (!path.startsWith("/")) {
+        throw new Error("Mount target does not start with /");
+      }
+      if (path.includes("//")) {
+        throw new Error("Mount target includes double /, please use a single one");
+      }
+    }
+    const mountOrder = Object.entries(config.fs).sort(
```

Comment:
Why do you need to specify the mount order? Does it need to be deterministic? If so, keep in mind this sort is not deterministic if you get two paths with the same number of slashes.

Reply:
I will add a comment explaining that:
```
    // We need to ensure that we mount devices from the top down,
    // If /asf would be mounted before / then the /asf mount wouldn't be
    // the parent of the / mount, breaking the system.
    // We only care to sort the mount targets from top down
    // This logic was omitted from the initrd cause it's easier to do it here
```